### PR TITLE
person search by date should return results in a timely manner

### PR DESCRIPTION
**Pull request by OriHoch**
* fix persons years search - make it use the appropriate fields + ensure only indexed fields are searched
* added scripts/mongo_create_index.py - creates an index to make the fsearch work faster
* fix migration process to get marriage_years
* added unit tests

#### post-deploy
* should run scripts/mongo_create_index.py - and see if it works on production DB and makes searches faster

### CODE REVIEW

File: bhs_api/fsearch.py [hunk 1]
```diff
@@ -29,13 +29,14 @@
                 }
 
 def _generate_year_range(year, fudge_factor=0):
-    maximum = int(str(year + fudge_factor) + '9999')
-    minimum = int(str(year - fudge_factor) + '0000')
+    maximum = int(str(year + fudge_factor))
+    minimum = int(str(year - fudge_factor))
```
Comment (OriHoch) [inline bhs_api/fsearch.py]: not sure why the additional numbers were added, currently we store the year as-is, as a number in mongo

File: bhs_api/fsearch.py [hunk 2]
```diff
@@ -153,6 +153,15 @@ def build_query(search_dict):
     return search_query
 
 
+def build_search_dict(**kwargs):
+    search_dict = {}
+    for key, value in kwargs.items():
+        search_dict[key] = value[0]
+        if not value[0]:
+            abort(400, "{} argument couldn't be empty".format(key))
+    return search_dict
```
Comment (OriHoch) [inline bhs_api/fsearch.py]: refactored out of fsearch function - to allow testing it directly

File: migration/family_trees.py [hunk 3]
```diff
@@ -90,23 +90,22 @@ def save(self, data, name):
...
-                        sex = e.gender)
+            node = dict(id=node_id, sex=e.gender)
             node['deceased'] = not is_living_person(e.deceased, e.birth_year)
             if not e.private:
                 node['name'] = e.name
                 if full and node['deceased']:
                     node['birth_year'] = e.birth_year
                     node['death_year'] = e.death_year
-                    node['marriage_years'] = e.marriage_years
+                    node['marriage_years'] = self.gedcom.marriage_years(e)
```
Comment (OriHoch) [inline migration/family_trees.py]: marriage_years is stored in the family section of GEDCOM - the gedcom.marriage_years function searchges the gedcom families to find the specific individual's marriage years

File: migration/family_trees.py [hunk 4]
```diff
@@ -90,23 +90,22 @@ def save(self, data, name):
         data['tree_file_id'] = self.meta['file_id']
         self.onsave(data, 'persons')
 
-    def flatten(self, nodes, full=False):
+    def flatten(self, node_or_nodes, full=False):
```
Comment (OriHoch) [inline migration/family_trees.py]: I refactored the function a bit to make it clearer that it allows either single or multiple nodes, before it was a bit unclear (to me..)

File: requirements.txt [hunk 5]
```diff
@@ -8,7 +8,7 @@ argparse==1.2.1
 boto==2.35.1
 flask-mongoengine==0.7.1
 mongoengine==0.10.0
-mongomock==3.2.1
+git+https://github.com/OriHoch/mongomock.git@support-for-operator-filtering-inside-list-fields#egg=mongomock
```
Comment (OriHoch) [inline requirements.txt]: my mongomock branch contains a fix for searching in list fields (for the marriage_years)

File: bhs_api/fsearch.py (hunk 1 above)
Comment (daonb) [inline bhs_api/fsearch.py]: probably something to do with the format of the old dates fields

File: bhs_api/fsearch.py [hunk 6]
```diff
@@ -29,13 +29,14 @@
...
-    minimum = int(str(year - fudge_factor) + '0000')
+    maximum = int(str(year + fudge_factor))
+    minimum = int(str(year - fudge_factor))
     return {'min': minimum, 'max': maximum}
 
 
 def build_query(search_dict):
-    ''' build a mongo search query based on the search_dict '''
+    ''' build a mongo search query based on the search_dict
+    '''
```
Comment (daonb) [inline bhs_api/fsearch.py]: > For one liner docstrings, please keep the closing """ on the same line.

PEP8

File: migration/family_trees.py (hunk 4 above)
Comment (daonb) [inline migration/family_trees.py]: :+1:

**Comment by daonb** [comment]
In mongo_create_index.py index for `Slug.En` is missing. Also, we should probably move all the create_index calls here and add the Slug indexes for all the collections

**Comment by OriHoch** [comment]
@daonb - this PR focuses on fixing #82 - regarding the persons search
I opened an issue to handle your comment regarding indexes - #111

**Pull request merged**